Source/VCGeneration/Split.cs

```diff
@@ -21,9 +20,50 @@ public record VCResult
     DateTime startTime,
     ProverInterface.Outcome outcome,
     TimeSpan runTime,
-    IEnumerable<AssertCmd> asserts,
+    int maxCounterExamples,
+    List<Counterexample> counterExamples,
+    List<AssertCmd> asserts,
     int resourceCount
-  );
+  ) {
+    public void ComputePerAssertOutcomes(out Dictionary<AssertCmd, ProverInterface.Outcome> perAssertOutcome,
+      out Dictionary<AssertCmd, Counterexample> perAssertCounterExamples) {
+      perAssertOutcome = new();
+      perAssertCounterExamples = new();
+      if (outcome == ProverInterface.Outcome.Valid) {
+        perAssertOutcome = asserts.ToDictionary(cmd => cmd, assertCmd => ProverInterface.Outcome.Valid);
+      } else {
+        foreach (var counterExample in counterExamples) {
+          // Only deal with the ocunter-examples that cover the asserts of this split.
+          AssertCmd underlyingAssert;
+          if (counterExample is AssertCounterexample assertCounterexample) {
+            underlyingAssert = assertCounterexample.FailingAssert;
+          } else if (counterExample is CallCounterexample callCounterexample) {
+            underlyingAssert = callCounterexample.FailingAssert;
+          } else if (counterExample is ReturnCounterexample returnCounterexample) {
+            underlyingAssert = returnCounterexample.FailingAssert;
+          } else {
+            continue;
+          }
+
+          if (!asserts.Contains(underlyingAssert)) {
+            continue;
+          }
+
+          perAssertOutcome.Add(underlyingAssert, ProverInterface.Outcome.Invalid);
+          perAssertCounterExamples.Add(underlyingAssert, counterExample);
+        }
+
+        var remainingOutcome =
```

Comment:
This part is a bit confusing to me. I get line 58, but line 59 I'm not sure about. Shouldn't that be `ProverInterface.Outcome.Undetermined`?

Reply:
You have a point, let me explain the other one.
`ProverInterface.Outcome.Undetermined` was my first choice too originally, but then I figured that if we arrive at this point, then either `outcome != ProverInterface.Outcome.Invalid` or `counterExamples.Count >= maxCounterExamples`
Hence.
1. If `outcome != ProverInterface.Outcome.Invalid`, then because `outcome != ProverInterface.Outcome.Valid` as well (we are on the else branch), the remaining outcomes are either `TimeOut`, `OutOfMemory`, `OutOfResource` or `Undetermined` (I'm not sure what `Bounded` means), which are all more precise than just `Undetermined`, and thus it's nice to report such information at the assertion level.
2. Now if outcome == ProverInterface.Outcome.Invalid, then `counterExamples.Count >= maxCounterExamples`, it means we would assign `ProverInterface.Outcome.Invalid` to the remaining assertions, which is arguably incorrect, so let me change this particular case to `Undetermined` instead.